### PR TITLE
Support .msi in bouncerscript

### PR DESCRIPTION
**Pull request by Callek**
This is not yet ready for review, but the patch set will support .msi for Nightly and releases

### CODE REVIEW

File: bouncerscript/task.py [hunk 1]
```diff
@@ -96,9 +96,10 @@ def check_product_names_match_aliases(context):
 
 def check_product_names_match_nightly_locations(context):
     """Double check that nightly products are as expected"""
-    products = context.task["payload"]["bouncer_products"]
-    if sorted(products) != sorted(BOUNCER_PATH_REGEXES_PER_PRODUCT.keys()):
-        raise TaskVerificationError("Products {} don't correspond to nightly ones".format(products))
+    products = set(context.task["payload"]["bouncer_products"])
+    valid_products = set(BOUNCER_PATH_REGEXES_PER_PRODUCT.keys())
+    if not products.issubset(valid_products):
```
Comment (MihaiTabara) [inline bouncerscript/task.py]: I'm not sure this is correct. Thing is, whenever we want to update the nightly products, we need to update all of them at once, not a subset of them. So if say, there's an in-tree misconfiguration/malicious configuration and, say, only two of the nightlies products are in the payload, that means we'll end up with various nightly products serving different versions. This should be strict comparison, either all of them or none.

On the other hand, I understand why you approached it this way. MSi is windows specific and hence won't always be there. But we only have one `bouncer-locations` job per nightly release so I don't see why not keep as it was and simply enrich the list of products with `firefox-msi-latest-ssl` as well?

File: bouncerscript/task.py [hunk 2]
```diff
@@ -97,7 +97,10 @@ def check_product_names_match_aliases(context):
 def check_product_names_match_nightly_locations(context):
     """Double check that nightly products are as expected"""
     products = context.task["payload"]["bouncer_products"]
-    if sorted(products) != sorted(BOUNCER_PATH_REGEXES_PER_PRODUCT.keys()):
+    valid_sets = []
+    for product_set in BOUNCER_PATH_REGEXES_PER_PRODUCT:
+        valid_sets.append(sorted(product_set.keys()))
+    if not sorted(products) in valid_sets:
```
Comment (MihaiTabara) [inline bouncerscript/task.py]: Nit: For some reason it seems more explicit to me to be `if sorted(products) not in valid_sets` but either is fine.

File: bouncerscript/test/test_task.py [hunk 3]
```diff
@@ -619,6 +619,15 @@ def test_check_path_matches_destination(product_name, path, raises):
         "firefox-nightly-latest-l10n",
         "firefox-nightly-latest-l10n-ssl"
     ], False
+), (
```
Comment (MihaiTabara) [inline bouncerscript/test/test_task.py]: Maybe worth adding another test with the following:
```
`firefox-nightly-msi-latest-ssl`
`firefox-nightly-msi-latest-l10n-ssl`
```
with `raises` set to `True`. 

AND

 ```
`firefox-nightly-msi-latest-ssl`
`firefox-nightly-msi-latest-l10n-ssl`
`firefox-nightly-latest`
```
with `raises` set to `True`

AND 

 ```
`firefox-nightly-msi-latest-ssl`
`firefox-nightly-latest`
```
with `raises` set to `True`

File: bouncerscript/test/test_task.py [hunk 4]
```diff
@@ -692,6 +701,26 @@ def test_check_version_matches_nightly_regex(version, raises):
     'firefox-nightly-latest',
     '/firefox/nightly/latest-mozilla-central-l10n/firefox-63.0a1.:lang.linux-x86_64.tar.bz2',
     False
+), (
```
Comment (MihaiTabara) [inline bouncerscript/test/test_task.py]: Nice 👍

File: bouncerscript/task.py [hunk 5]
```diff
@@ -109,9 +112,16 @@ def check_locations_match(locations, product_config):
...
-               BOUNCER_PATH_REGEXES_PER_PRODUCT[product_name],
-               fullmatch=True) is None:
+    match = None
+    for product_regex in BOUNCER_PATH_REGEXES_PER_PRODUCT:
+        if product_name not in product_regex:
+            continue
+        match = matches(path,
+                        product_regex[product_name],
+                        fullmatch=True)
+        if match:
```
Comment (MihaiTabara) [inline bouncerscript/task.py]: 👍 for breaking here, otherwise we would check some of the paths twice.

File: bouncerscript/constants.py [hunk 6]
```diff
@@ -32,7 +32,7 @@
     'thunderbird': r'^(/thunderbird/releases/.*?/(update/)?(?:linux-i686|linux-x86_64|mac|win32|win64)/\:lang/(?:thunderbird|Thunderbird).*\.(?:bz2|dmg|exe|mar))$',
 }
 
-BOUNCER_PATH_REGEXES_PER_PRODUCT = {
+_BOUNCER_PATH_REGEXES_PER_PRODUCT_0 = {
```
Comment (MihaiTabara) [inline bouncerscript/constants.py]: Nit: I'm tempted to rename this to `_BOUNCER_PATH_REGEXES_PER_PRODUCT_DEFAULT` while the one containing the `msi` to `_BOUNCER_PATH_REGEXES_PER_PRODUCT_MSI`. What do you think?

**Comment by Callek** [comment]
@MihaiTabara So I think I've made all the necessary changes here, can you recommend how you would like me to test this and the queue in https://bugzilla.mozilla.org/show_bug.cgi?id=1475512 before deploying?

File: bouncerscript/constants.py [hunk 7]
```diff
@@ -10,29 +10,34 @@
     'firefox-devedition-stub': r'^Devedition-\d+\.0b\d+-stub$',
     'firefox-devedition-latest': r'^Devedition-\d+\.0b\d+$',
     'firefox-devedition-latest-ssl': r'^Devedition-\d+\.0b\d+-SSL$',
+    'firefox-devedition-msi-latest-ssl': r'^Devedition-\d+\.0b\d+-msi-SSL$',
```
Comment (MihaiTabara) [inline bouncerscript/constants.py]: We need to add all these newly-added entries into bouncer.

**Comment by MihaiTabara** [comment]
@Callek Easiest is to cherry pick the in-tree changeset to maple and pin bouncerscript-dev to encompass these change. Once we're confirmed things are working as expected, we can merge this and land to `central`. I can help with bouncerscript + bouncer parts.

**Pull request merged**